plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.hs

```diff
@@ -348,6 +451,67 @@ extKCS k KCS{..} = KCS{ kindOf = kindOf' }
 
 -- ** Type checking
 
+instance Check TypeBuiltinG TermConstantG where
+  check TyByteStringG (TmByteStringG _) = true
+  check TyIntegerG    (TmIntegerG    _) = true
+  check TyStringG     (TmStringG     _) = true
+  check TyBoolG       (TmBoolG       _) = true
+  check TyCharG       (TmCharG       _) = true
+  check TyUnitG       (TmUnitG       _) = true
+  check _             _                 = false
```

Comment:
That definition would be trivial if the universes stuff was used everywhere. It would be just "check that two type tags are equal".

Reply:
I think the benefit of reusing syntax for type and term constants relies on automatically generating the types of builtins otherwise it's painful to write out long names of things.